src/main/scala/is/hail/methods/LinearMixedRegressionDist.scala

```diff
@@ -0,0 +1,216 @@
+            .orderedRepartitionBy(vds1.rdd.orderedPartitioner)
+
+          vds1.rdd.orderedLeftJoinDistinct(projG).asOrderedRDD.mapPartitions({ it =>
+            val missingSamples = new ArrayBuilder[Int]
+
+            // columns are genotype vectors
+            var projX: DenseMatrix[Double] = null
+
+            it.grouped(blockSize)
+              .flatMap(git => {
+                val block = git.toArray
+                val blockLength = block.length
+
+                if (projX == null || projX.cols != blockLength)
+                  projX = new DenseMatrix[Double](nEigs, blockLength)
+
+                var i = 0
+                while (i < blockLength) {
+                  val (_, ((_, _), Some(px))) = block(i)
+
+                  projX(::, i) := px
+
+                  i += 1
+                }
+
+                val annotations = scalarLMMBc.value.likelihoodRatioTestBlock(projX)
+
+                (block, annotations).zipped.map { case ((v, ((va, gs), _)), a) => (v, (inserter(va, a), gs)) }
+              })
+          }, preservesPartitioning = true)
+        } else {
+          val scalarLMM = LowRankScalarLMM(lmmConstants, diagLMM.delta, diagLMM.logNullS2, useML)
+          val scalarLMMBc = sc.broadcast(scalarLMM)
+
+          val projG = (allG.toBlockMatrixDense() * Ut.transpose)
+            .toIndexedRowMatrix()
+            .rows
+            .map { case IndexedRow(i, px) => (variantsBc.value(i.toInt), DenseVector(px.toArray)) }
+            .orderedRepartitionBy(vds1.rdd.orderedPartitioner)
+
+          vds1.rdd.orderedLeftJoinDistinct(projG).asOrderedRDD.mapPartitions({ it =>
+            val sclr = scalarLMMBc.value
+
+            val r2 = 1 to c
+
+            val CtC = DenseMatrix.zeros[Double](c + 1, c + 1)
+            CtC(r2, r2) := sclr.con.CtC
+
+            val UtC = DenseMatrix.zeros[Double](nEigs, c + 1)
+            UtC(::, r2) := sclr.Utcov
+
+            val Cty = DenseVector.zeros[Double](c + 1)
+            Cty(r2) := sclr.con.Cty
+
+            val CzC = DenseMatrix.zeros[Double](c + 1, c + 1)
+            CzC(r2, r2) := sclr.UtcovZUtcov
+
+            val missingSamples = new ArrayBuilder[Int]
+            val x = DenseVector.zeros[Double](n)
+
+            it.map { case (v, ((va, gs), Some(px))) =>
+              if (useDosages)
+                RegressionUtils.dosages(x, gs, completeSampleIndexBc.value, missingSamples)
+              else
+                x := RegressionUtils.hardCalls(gs, n, sampleMaskBc.value) // No special treatment of constant
+
+              val annotation = sclr.likelihoodRatioTestLowRank(x, px, CtC, UtC, Cty, CzC)
+
+              (v, (inserter(va, annotation), gs))
+            }
+          }, preservesPartitioning = true)
+        }
+
+      vds1.unpersist()
+
+      vds1.copy(
+        rdd = newRDD.asOrderedRDD,
+        vaSignature = newVAS)
+    } else
+      vds1
+
+    vds2
+  }
+}
+
+object Helper {
```

Comment:
If you are going to keep this named Helper, it cannot have global scope the way it does here. It should probably be inside the `LinearMixedRegressionDist` object anyway, regardless of name.